src/aio.rs

```diff
@@ -0,0 +1,325 @@
+        let rc = unsafe {
+            libc::syscall(
+                libc::SYS_io_getevents,
+                self.0,
+                min_nr,
+                events.len() as c_long,
+                events.as_mut_ptr(),
+                to,
+            ) as c_int
+        };
+        if rc < 0 {
+            Err(Error::last_os_error())
+        } else {
+            Ok(rc as usize)
+        }
+    }
+}
+
+impl Drop for IoContext {
+    fn drop(&mut self) {
+        if self.0 != 0 {
+            // It's safe because the context is created by us.
+            let _ = unsafe { libc::syscall(libc::SYS_io_destroy, self.0) as c_int };
+        }
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use std::fs::File;
+    use std::os::unix::io::AsRawFd;
+
+    #[test]
+    fn new_context() {
+        let _ = IoContext::new(0).unwrap_err();
+    }
+
+    #[test]
+    fn cancel_request() {
+        let file = File::open("/dev/zero").unwrap();
+
+        let ctx = IoContext::new(128).unwrap();
+        let mut buf: [u8; 16384] = unsafe { std::mem::uninitialized() };
+        let iocbs = [&mut IoControlBlock {
+            aio_fildes: file.as_raw_fd() as u32,
+            aio_lio_opcode: IOCB_CMD_PREAD as u16,
+            aio_buf: buf.as_mut_ptr() as u64,
+            aio_nbytes: buf.len() as u64,
+            ..Default::default()
+        }];
+
+        let mut rc = ctx.submit(&iocbs).unwrap();
+        assert_eq!(rc, 1);
+
+        let mut result = Default::default();
+        let err = ctx
+            .cancel(iocbs[0], &mut result)
+            .unwrap_err()
+            .raw_os_error()
+            .unwrap();
+        assert_eq!(err, libc::EINVAL);
+
+        let mut events = [unsafe { std::mem::uninitialized::<IoEvent>() }];
+        rc = ctx.get_events(1, &mut events, None).unwrap();
+        assert_eq!(rc, 1);
+        assert!(events[0].res > 0);
+    }
+
+    #[test]
+    fn read_zero() {
+        let file = File::open("/dev/zero").unwrap();
+
+        let ctx = IoContext::new(128).unwrap();
+        let mut buf: [u8; 4096] = unsafe { std::mem::uninitialized() };
+        let iocbs = [
+            &mut IoControlBlock {
+                aio_fildes: file.as_raw_fd() as u32,
+                aio_lio_opcode: IOCB_CMD_PREAD as u16,
+                aio_buf: buf.as_mut_ptr() as u64,
+                aio_nbytes: buf.len() as u64,
+                ..Default::default()
+            },
+            &mut IoControlBlock {
+                aio_fildes: file.as_raw_fd() as u32,
+                aio_lio_opcode: IOCB_CMD_PREAD as u16,
+                aio_buf: buf.as_mut_ptr() as u64,
+                aio_nbytes: buf.len() as u64,
+                ..Default::default()
+            },
+        ];
+
+        let mut rc = ctx.submit(&iocbs[..]).unwrap();
+        assert_eq!(rc, 2);
+
+        let mut events = [unsafe { std::mem::uninitialized::<IoEvent>() }];
+        rc = ctx.get_events(1, &mut events, None).unwrap();
```

Comment:
Not sure how can we ensure the event is in completion queue?

Reply:
we are reading /dev/zero, it should not block at all.

Reply:
It would be nice to copy the test as examples for those functions so that public users can know how to use it explicitly.

Reply:
Fixed:)